homeassistant/components/evohome/__init__.py

```diff
@@ -221,9 +272,116 @@ def _handle_exception(err) -> bool:
         broker.update, config[DOMAIN][CONF_SCAN_INTERVAL]
     )
 
+    setup_service_functions(hass, broker)
+
     return True
 
 
+def setup_service_functions(hass: HomeAssistantType, broker):
+    """Set up the service functions."""
+
+    def _clean_dt(dtm):
+        if dtm is not None:
+            format_str = "%Y-%m-%d %H:%M:00"  # round down to the nearest minute
+            dtm = dt.strptime(dt.strftime(dtm, format_str), format_str)
+            if dtm < dt.now():
+                dtm += timedelta(days=1)
+        return dtm
+
+    async def force_refresh(call):
+        """Obtain the latest state data via the vendor's RESTful API."""
+        await broker.update()
+
+    async def set_system_mode(call):
+        """Set the system mode."""
+        if call.service == SVC_SET_SYSTEM_MODE:
+            mode = call.data[ATTR_SYSTEM_MODE]
+        else:
+            mode = EVO_RESET
+
+        tcs = broker.tcs
+        try:
+            attrs = [m for m in tcs.allowedSystemModes if m["systemMode"] == mode][0]
+        except IndexError:
+            raise KeyError(f"'{mode}' mode is not supported by this system")
+
+        until = None
+        # the following two attrs are mutually exclusive
+        if ATTR_DURATION_DAYS in call.data:
+            if attrs.get("timingResolution") != "1.00:00:00":
+                raise TypeError(f"'{mode}' mode does not support days, only hours")
+
+            until = dt.combine(dt.now().date(), dt.min.time())
+            until += timedelta(days=call.data[ATTR_DURATION_DAYS])
+
+        if ATTR_DURATION_HOURS in call.data:
+            if attrs.get("timingResolution") != "01:00:00":
+                raise TypeError(f"'{mode}' mode does not support hours, only days")
+            until = dt.now() + timedelta(hours=call.data[ATTR_DURATION_HOURS])
+
+        await broker.call_client_api(
+            tcs._set_status(mode, _clean_dt(until))  # pylint: disable=protected-access
+        )
+
+    async def set_zone_setpoint(call):
+        """Set the system mode."""
+        entity_id = call.data[ATTR_ENTITY_ID]
+
+        try:
+            unique_id = hass.data["entity_registry"].entities[entity_id].unique_id
+            zone_entity = broker.entities[unique_id]
```

Comment:
We're not allowed to access the entity outside the platform directly. We should use the dispatch helper to signal the correct entity method from here.

See the tuya integration for an example.

Reply:
Done